### PR TITLE
SONARJAVA-2158 S2974 Don't raise issue when class is extended

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-checks/src/test/files/checks/FinalClassCheck.java [hunk 1]
```diff
@@ -49,3 +49,23 @@ public static int magic(){
 
 public class NoConstructorClass { // Compliant, implicit constructor has package visibility
 }
+
+public abstract class AbstractClass {  // Compliant, constructor is used in
```
Comment (benzonico) [inline java-checks/src/test/files/checks/FinalClassCheck.java]: This test lacks the case of an anonymous class : new AbstractClass("") {}  

I am  not sure if your code will work properly with this case or not so it deserves to be tested at least.

File: java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java [hunk 2]
```diff
@@ -41,12 +42,23 @@
...
+    if (hasOnlyPrivateConstructors(classTree) && !isExtended(classTree) && !ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.FINAL)) {
       reportIssue(classTree.simpleName(), "Make this class \"final\" or add a public constructor.");
     }
   }
 
+  private static boolean isExtended(ClassTree classTree) {
+    List<IdentifierTree> usages = classTree.symbol().usages();
+    return usages.stream().anyMatch(usage -> {
+      Tree parent = usage.parent();
+      return parent != null && parent.is(Kind.CLASS);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java]: This won't work if fully qualified name is used in the extend clause.
```
class Foo extends my.pack.age.AbstractClass {} 
```
I would more go for a subvisitor looking for type of extend clause and anonymous constructor.

File: java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java [hunk 3]
```diff
@@ -23,11 +23,16 @@
 import org.sonar.check.Rule;
 import org.sonar.java.model.ModifiersUtils;
 import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
 import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java]: MINOR Remove this unused import 'org.sonar.plugins.java.api.tree.IdentifierTree'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java [hunk 4]
```diff
@@ -60,4 +74,30 @@ private static boolean hasOnlyPrivateConstructors(ClassTree classTree) {
...
+
+    IsExtendedVisitor(Type type) {
+      this.type = type;
+    }
+
+    @Override
+    public void visitClass(ClassTree tree) {
+      TypeTree superClass = tree.superClass();
+      if (superClass != null && superClass.symbolType() == type) {
+        isExtended = true;
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java]: once set to true, you can cut the visit of the subtree.

File: java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java [hunk 5]
```diff
@@ -60,4 +74,30 @@ private static boolean hasOnlyPrivateConstructors(ClassTree classTree) {
...
+      }
+      super.visitClass(tree);
+    }
+
+    @Override
+    public void visitNewClass(NewClassTree tree) {
+      if (tree.classBody() != null && tree.symbolType().isSubtypeOf(type)) {
+        isExtended = true;
+      }
+      super.visitNewClass(tree);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/FinalClassCheck.java]: same here, when isExtended is set to true, no need to visit the subtree.

**Pull request merged**